### DISCUSSION TITLE
Discrete mapper improvements

### DISCUSSION THREAD

**Discussion by maxkfranz** [cytoscape/cytoscape-explore#99] [Ideas]
Here's an idea to improve the discrete mapper UI.  Currently, there are two issues:

- (A) The mapper UI shows a lot of options when the the set of values for the selected attribute is high.
- (B) The issue re. 'defaults'.

The idea is to make the list of values editable:

- (1) Each line in the discrete mapper UI shows an 'x' beside it so it can be removed.
- (2) There's a plus button to add a new line.
- (3) Whenever not all of the values each have a line, there is an 'other' entry at the end.  The 'other' entry is not shown to the user otherwise, since it would have no net effect to change it when all the real values have mappings set.
- (4) The initial set of discrete mappings is smart, based on the size of the set of values.  If large, then you start off with just the plus button and 'other'.  If small, then you start off with all the entries listed with lines.
- (5) When clicking the plus button, you just use a dropdown to select the attribute you want.

It would look something like this:

```
x   [ red   ]   Some value
x   [ blue  ]   Another value
x   [ green ]   Other

(Add a entry)
```

The main benefits are:

- You always see the full list when it's a reasonable size to map.  You shouldn't be using a discrete mapper to set individual colours for 200 distinct score values, for instance.  But if you have just 6 score values, then it's reasonable.
- You don't get overwhelmed by a huge, difficult to edit list.
- The 'defaults' issue is resolved by 'other', yet the user doesn't get the added complexity of 'defaults' semantics.  The mental model for users is 'all the other values', not a hierarchical set of rules with defaults.
- This is more in line with the legends idea.  The discrete mapper UI looks pretty well like a legend would, which means it would fit well with future legend UIs and it would fit naturally with the user's mental model.
- In terms of conversion to CX etc., the 'default' value doesn't matter unless the user hasn't set all the values.
  - If the user has set all values: The default doesn't matter.  You could send a random value if you really want.
  - If the user has set only some values:  Send 'other' as the default value.

@mikekucera @chrtannus @d2fong

**Comment by mikekucera** [comment]
I agree with this idea, I was basically thinking the same thing :)

We could eventually take this idea even further. For example when you click the + to add an entry to the discrete mapping there could be options for ranges of values. For example the user could choose that all positive values are red, and all negative values are blue. In Cytoscape to achieve this you could use a discreet mapper and map every individual positive value to red, etc. Or you would have to create a continuous mapping with multiple data points. When really all the user wants is to say positive=red, negative=blue.

My issue with the 'default' option in cytoscape is that when the mapping has no gaps then changing the default does nothing, but its always displayed no matter what. In CE we could detect the case where the mapping has gaps and only show the picker for the default value when there are gaps. For discreet mappings just an extra row that says 'other' seems like a good way of presenting it. For the linear mapper its not as obvious. We could put an extra button next to the mapping and see how that goes.

**Comment by mikekucera** [comment]
The more I think about this the more I become unsure. Imagine the user really does want to map 20 discrete values to different colors. This idea will require a lot of extra clicking. For every data value they have to click the (Add an entry) button, then select the value, then select the color, times 20. With the current UI the data values are already all listed, the user just needs to select the colors.

Also, cytoscape desktop does have a nice feature for bootstrapping a large discrete mapping. After creating the mapping, if you right click any data value you can select a color palette, which completely populates the mapping. If the user doesn't care what the colors are, just that different values have different colors, then they can create a discrete mapping from a palette with only a few clicks.

<img width="1301" alt="Screen Shot 2021-11-17 at 4 13 30 PM" src="https://user-images.githubusercontent.com/5350528/142283474-3a9f0d50-2976-45dd-bf01-5269fc05f300.png">

**Comment by maxkfranz** [comment]
Discrete mappers are kludgey.  I don't think we should worry right now about making it too easy to make discrete mappings with lots of entries.  We shouldn't be encouraging the user to do that in the first place, in my opinion.  

Discrete mappings of more than 6 entries aren't typically good as visualisations -- and around 10 or 12 is really pushing it.  Would someone, other than the author, understand what's going on in a figure with 20 colours?  Could you even distinguish that many colours visually without making mistakes often?  

I don't want to make it easy for our users to create visually confusing visualisations.  If a reader of a journal saw an attractive, well-designed network made and cited with CE, they would be much more likely to use CE for their own articles than if they saw a confusing, messy figure.

That's why I still think selection-based styling would be best.  It would encapsulate everything that the existing style model can accommodate but in a much simpler way for the user.

Fo now, I think we should keep the discrete mapper UI but just make it smaller and simpler with good restrictions (e.g. don't allow too many entries).  Even if there isn't a hard cap, the friction is a good nudge.

**Comment by mikekucera** [comment]
I think you're probably right about 6+ entries being too much. I think if you could use a discreet mapper to map ranges of values and not just individual data values it could be more powerful. For example create a mapping where every value >1 is red, all others are blue. That could be done with a discreet mapper with two entries. You can do this in cytoscape desktop now, but only with a continuous mapper, and you have to get pretty familiar with the continuous mapping editor dialog.

I'll start working updating the UI for discreet mappings as shown in the first post, and we'll see how it goes.

**Comment by maxkfranz** [comment]
Yeah, I think your observation about the ranges is spot-on. I still that that when we’re exploring that level of detail that we should evaluate a selection-based mechanism for the UI. Your discrete mapper model with ranges is a specific case of a general selection mechanism: The user makes a selection for value>1 (probably with a search UI that would be useful for general selection anyway) and then can just set them blue the same way that they would normally do.

**Comment by mikekucera** [comment]
Pull request:  #106

**Comment by mikekucera** [comment]
We should discuss what to do about scrollbars that appear when the contents of the style panel fill the vertical space. They look pretty ugly in my opinion.
Is there a way to use CSS to style the scrollbars so they at least match the dark color scheme of the toolbar?

<img width="1020" alt="Screen Shot 2021-11-23 at 3 20 00 PM" src="https://user-images.githubusercontent.com/5350528/143099316-ca5e990f-1bf5-4790-a7f7-859ee1230fad.png">

**Reply by maxkfranz** [reply]
We could set it up like this:

If the user has set the preference to show inline scrollbars all the time in the Mac Preferences app, then we show an inline scrollbar styled to match our app.

If the user has the default preference (e.g. iOS scrollbars, modern scrollbars on Mac, GNOME, etc.), then the native scrollbars are used since modern scrollbars already look fine -- and native scrollbars are generally preferable where possible, in my opinion.
